src/seismometer/_api.py

```diff
@@ -176,86 +183,126 @@ def fairness_audit(metric_list: Optional[list[str]] = None, fairness_threshold=1
     sensitive_groups = sg.cohort_cols
     metric_list = metric_list or ["tpr", "fpr", "pprev"]
 
-    df = pdh.event_score(
-        sg.data(),
-        sg.entity_keys,
-        score=sg.output,
-        ref_event=sg.predict_time,
-        aggregation_method=sg.event_aggregation_method(sg.target),
-    )[[sg.target, sg.output] + sensitive_groups]
-
-    display_fairness_audit(
-        df, sensitive_groups, sg.output, sg.target, sg.thresholds[0], metric_list, fairness_threshold
+    return generate_fairness_audit(
+        sensitive_groups,
+        sg.target,
+        sg.output,
+        sg.thresholds[0],
+        per_context=True,
+        metric_list=metric_list,
+        fairness_threshold=fairness_threshold,
     )
 
 
-# endregion
-
-# region notebook IPWidgets # TODO - move to .controls submodule
+@export
+def generate_fairness_audit(
```

Comment:
Q: when are docstrings needed and when are they not?

Reply:
We should try to include them for anything we expect to be a public facing API.  The most common exception should be private methods (prefixed with an underscore) 

A:  This should have a docstring :-)

Reply:
FINC